pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/version"
+)
+
+const (
+	// how many resolved timestamps to wait for before considering the
+	// system to be working as intended.
+	requestedResolved = 20
+)
+
+var (
+	// the CDC target (table). We're running the tpcc workload in this
+	// test; verifying that the changefeed on the `warehouse` table
+	// works is sufficient for our purposes
+	target = "tpcc.warehouse"
+)
+
+func registerCDCMixedVersions(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:            "cdc/mixed-versions",
+		Owner:           registry.OwnerTestEng,
+		Cluster:         r.MakeClusterSpec(5),
+		Timeout:         30 * time.Minute,
+		RequiresLicense: true,
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			runCDCMixedVersions(ctx, t, c, *t.BuildVersion())
+		},
+	})
+}
+
+// cdcMixedVersionTester implements mixed-version/upgrade testing for
+// CDC. It knows how to set up the cluster to run Kafka, monitor
+// events, and run validations that ensure changefeeds work during and
+// after upgrade.
+type cdcMixedVersionTester struct {
+	ctx           context.Context
+	crdbNodes     option.NodeListOption
+	workloadNodes option.NodeListOption
+	kafkaNodes    option.NodeListOption
```

Comment:
Do we want to restrict this to only work with kafka?

Reply:
I'm open to supporting other sinks. Do we have tests that pull changes from other sinks at the moment? I can't seem to find anything, but maybe I missed something.